### PR TITLE
Fee Rate Api Support

### LINKED ISSUE #1163: Implement ability to fetch fee rates from APIs
Create a `FeeFetcher` interface in the core project, and then implement a few fee provider APIs and then allow a fee provider to be specific in our config file

### PR DESCRIPTION
**Pull request by benthecarman**
Closes #1163 

Preliminary api support for fetching fee rates. Added https://bitcoiner.live/ to start off with because it had a simple api.

### CODE REVIEW

**Review by Christewart — COMMENTED** [review]
So I think we should actually breakout the fee providers into a separate module. I don't want to add a bunch of dependencies into the wallet project. Then the wallet can just take a `FeeRateApi` as a parameter to the `WalletApi`

I also think the `FeeRateApi` should be moved into the `core` project so that it is available for everyone to implement. The `FeeRateApi` should not have any implemented methods

File: app/server/src/main/scala/org/bitcoins/server/WalletRoutes.scala [hunk 1]
```diff
@@ -97,13 +97,12 @@ case class WalletRoutes(wallet: WalletApi, node: Node)(
...
-            val feeRate =
-              satoshisPerVirtualByteOpt.getOrElse(SatoshisPerByte(100.satoshis))
-            wallet.sendToAddress(address, bitcoins, feeRate).map { tx =>
-              node.broadcastTransaction(tx)
-              Server.httpSuccess(tx.txIdBE)
-            }
+            wallet
+              .sendToAddress(address, bitcoins, satoshisPerVirtualByteOpt)
+              .map { tx =>
+                node.broadcastTransaction(tx)
```
Comment (Christewart) [inline app/server/src/main/scala/org/bitcoins/server/WalletRoutes.scala]: We should probably `.flatMap { _ => Server.httpSuccess(tx.txIdBe) }`

File: app/server/src/main/scala/org/bitcoins/server/WalletRoutes.scala [hunk 2]
```diff
@@ -156,14 +150,14 @@ case class WalletRoutes(wallet: WalletApi, node: Node)(
...
-            val feeRate =
-              satoshisPerVirtualByteOpt.getOrElse(SatoshisPerByte(100.satoshis))
-            wallet.makeOpReturnCommitment(message, hashMessage, feeRate).map {
-              tx =>
+            wallet
+              .makeOpReturnCommitment(message,
+                                      hashMessage,
+                                      satoshisPerVirtualByteOpt)
+              .map { tx =>
                 node.broadcastTransaction(tx)
```
Comment (Christewart) [inline app/server/src/main/scala/org/bitcoins/server/WalletRoutes.scala]: same here

File: fee-provider/src/main/scala/org/bitcoins/feeprovider/BitcoinerLiveFeeRateProvider.scala [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+case class BitcoinerLiveFeeRateProvider(minutes: Int)(
+    implicit override val system: ActorSystem)
+    extends CachedHttpFeeRateApi {
+  private val bitcoinerLiveValidMinutes =
+    Vector(30, 60, 120, 180, 360, 720, 1440)
+  require(
+    bitcoinerLiveValidMinutes.contains(minutes),
+    s"$minutes is not a valid selection, must be from $bitcoinerLiveValidMinutes")
+
+  override def uri: Uri =
```
Comment (Christewart) [inline fee-provider/src/main/scala/org/bitcoins/feeprovider/BitcoinerLiveFeeRateProvider.scala]: nit: `override val`

File: fee-provider/src/main/scala/org/bitcoins/feeprovider/HttpFeeRateApi.scala [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+      )
+  }
+}
+
+abstract class HttpFeeRateApi extends FeeRateApi {
+  implicit val system: ActorSystem
+
+  def uri: Uri
+
+  def converter: String => Try[FeeUnit]
```
Comment (Christewart) [inline fee-provider/src/main/scala/org/bitcoins/feeprovider/HttpFeeRateApi.scala]: lt's just make this a normal method 

```scala
def converter(str: String): Try[FeeUnit] = ...
```

File: fee-provider/src/main/scala/org/bitcoins/feeprovider/HttpFeeRateApi.scala [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+      .makeApiCall(uri)
+      .flatMap(ret => Future.fromTry(converter(ret)))(system.dispatcher)
+  }
+}
+
+abstract class CachedHttpFeeRateApi extends HttpFeeRateApi {
+
+  private var cachedFeeRateOpt: Option[(FeeUnit, Instant)] = None
+
+  val cacheDuration: Duration = Duration.ofMinutes(5)
```
Comment (Christewart) [inline fee-provider/src/main/scala/org/bitcoins/feeprovider/HttpFeeRateApi.scala]: Probably should be private, and also configurable in a 
```
bitcoin-s { 
  fee-provider { 
    cache-duration = 5.minutes 
  }
}
```

File: wallet/src/main/scala/org/bitcoins/wallet/api/WalletApi.scala [hunk 6]
```diff
@@ -444,10 +446,27 @@ trait WalletApi extends WalletLogger {
       outPoints: Vector[TransactionOutPoint],
       address: BitcoinAddress,
       amount: CurrencyUnit,
-      feeRate: FeeUnit): Future[Transaction] = {
+      feeRateOpt: Option[FeeUnit],
+      fromAccount: AccountDb): Future[Transaction] = {
+    for {
+      feeRate <- feeRateOpt match {
```
Comment (Christewart) [inline wallet/src/main/scala/org/bitcoins/wallet/api/WalletApi.scala]: Let's refactor this `feeRateOpt match { ... } ` pattern into a helper method somewhere

**Review by Christewart — COMMENTED** [review]
It looks like ther is no unit tests with `BitcoinLiveFeeRateProvider`. Also like I said online it would be nice to have a dynamic fee test inside of `WalletIntegrationTest`.

File: fee-provider/src/main/scala/org/bitcoins/feeprovider/HttpFeeRateApi.scala (hunk 5 above)
Comment (benthecarman) [reply fee-provider/src/main/scala/org/bitcoins/feeprovider/HttpFeeRateApi.scala]: I think having it public is better, it allows other people using the library to set their own `cacheDuration`

**Pull request merged**